architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+From the outset, then, when _just_ considering a simple environment and the concerns of the CRD developer, we can see a number of difficult problems. Now we can add to this the usage patterns that are in practice today.
+
+In Kubernetes, the following CRD patterns have emerged:
+
+- Generic CRD: A single CRD "generalizes" a description of something, and any programs that need that generalized object can use that CRD. The CRD is thus a shared resource with no single controller.
+- CRD and Singleton Controller: A specific application defines a specific CRD that only it uses. Only one controller uses that CRD.
+- Common CRD: Two or more tightly coupled programs use a CRD to update information that is shared between the programs. Microservice architectures sometimes use this approach. (ConfigMaps and Secrets are also used this way, and the documentation for CRDs even suggests this as an alternative to ConfigMaps and Secrets in this capacity)
+- CRD and per-namespace controllers: One CRD is used globally, but each namespace that uses this CRD has its own controller which watches only that namespace. This strategy is done to reduce the cluster-level permissions the controller must have.
+
+As a special case, some controllers install and manage their own CRDs. For the most part, we won't deal with these here, as these do not pose a problem for Helm.
+
+The following facts about CRDs should be kept in mind throughout this paper:
+
+- CRDs are mutable. An operator can update versions, schemas, names, etc. on a CRD _ad hoc_. The API server will not enforce restrictions like it does on Deployments or other objects.
+- When a CR (CRD instance) is written to Kubernetes (on an update, for example), it will be _rewritten_ to the version that the CRD has marked as default. This means that backward compatibility can break merely by updating the default version on a CRD object. [See this section of the docs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects), which reveals a few other edge cases.
+    - During a read operation, Kubernetes can rewrite an object version without rewriting the body of the CR. So you can get a resource that is marked with a version that it actually is not.
+    - On a write operation, a version field may be permanently  rewritten to a version other than the version given in the object to be written, but the body is not updated
```

Comment:
Not sure what you are challenging. Did you test this out?

Reply:
Not deliberately, but thinking back, I believe I have actually been in the situation where an updated CRD meant that an existing object could not be read-write roundtripped because a field had been renamed, and the stored version no longer met the schema.

That's not incompatible with this paragraph, but I think it could be clearer, as what you end up with is a storage-version which _is_ correct for the stored object, but is not the version of the object you wrote. The implication I get from this paragraph now is that you can end up with _invalid_ objects stored if you write a CR with the non-default version.